src/visitors/kinetic_block_visitor.cpp

```diff
@@ -478,6 +478,162 @@ void KineticBlockVisitor::visit_kinetic_block(ast::KineticBlock& node) {
+        auto name = std::make_shared<ast::Name>(
+            std::make_shared<ast::String>(fmt::format("{}{}_", stem, n_equations)));
+        local_names.push_back(name);
+
+        return name;
+    }
+
+  public:
+    std::shared_ptr<ast::Name> generate_forward_rate_name() {
+        auto kf = generate_local_name(kf_stem);
+        ++n_equations;
+
+        return kf;
+    }
+
+    std::pair<std::shared_ptr<ast::Name>, std::shared_ptr<ast::Name>> generate_rate_names() {
+        auto kf = generate_local_name(kf_stem);
+        auto kb = generate_local_name(kb_stem);
+        ++n_equations;
+
+        return {kf, kb};
+    }
+
+    std::shared_ptr<ast::Name> generate_source_name() {
+        auto source = generate_local_name(source_stem);
+        ++n_equations;
+
+        return source;
+    }
+
+    ast::LocalVarVector get_local_variables() {
+        ast::LocalVarVector locals;
+        for (const auto& name: local_names) {
+            locals.push_back(std::make_shared<ast::LocalVar>(name));
+        }
+        return locals;
+    }
+
+  private:
+    size_t n_equations = 0;
+    std::vector<std::shared_ptr<ast::Name>> local_names;
+};
+
+class LocalizeKineticRatesVisitor: public AstVisitor {
+  public:
+    LocalRateNames local_names;
+
+  private:
+    std::shared_ptr<ast::ExpressionStatement> localize_expression(
+        const std::shared_ptr<ast::Name>& name,
+        const std::shared_ptr<ast::Expression>& expression) {
+        auto local = std::make_shared<ast::BinaryExpression>(name,
+                                                             ast::BinaryOperator(ast::BOP_ASSIGN),
+                                                             expression);
+        return std::make_shared<ast::ExpressionStatement>(local);
+    }
+
+    template <class Node>
+    std::shared_ptr<Node> clone(const Node& node) {
+        return std::shared_ptr<Node>(node.clone());
+    }
+
+  public:
+
+    void visit_kinetic_block(ast::KineticBlock& node) {
+        // process the statement block first.
+        node.visit_children(*this);
+
+        // We now know the names of the created LOCAL variables. If a LOCAL
+        // block exists, append to that, otherwise create a new one.
+        auto locals = local_names.get_local_variables();
+
+        auto stmt_block = node.get_statement_block();
+        const auto& stmts = stmt_block->get_statements();
+
+        auto it = std::find_if(stmts.cbegin(), stmts.cend(), [](const auto& stmt) {
+            return stmt->is_local_list_statement();
+        });
+
+        if (it != stmts.cend()) {
+            auto local_block = std::dynamic_pointer_cast<ast::LocalListStatement>(*it);
+            for (const auto& local: locals) {
+                local_block->emplace_back_local_var(local);
+            }
+        } else {
+            auto local_stmt = std::make_shared<ast::LocalListStatement>(locals);
+            stmt_block->insert_statement(stmts.cbegin(), local_stmt);
+        }
```

Comment:
could you check if we can use [add_local_variable()](https://github.com/BlueBrain/nmodl/blob/735b11dba9a494a5139be2ad5b82ca15e12859db/src/visitors/visitor_utils.cpp#L111) from visitor_utils?

Reply:
done.